### PR TITLE
Deprecate matchall

### LINKED ISSUE #26049: Unexpected behaviour in `matchall`
For example the code below returns the captured `()` `SubString` with the rest of the expression rather than just the capture request.

```
matchall(r"<tag>(.*?)</tag>", "<tag>I</tag> <tag>like</tag> <tag>writing</tag> <tag>code</tag>")
# 4-element Array{SubString{String},1}:
#  "<tag>I</tag>"      
#  "<tag>like</tag>"   
#  "<tag>writing</tag>"
#  "<tag>code</tag>"
```

The expected behaviour should return:
```
"I"
"like"
"writing"
"code"
```
Or maybe:

```
["I"]
["like"]
["writing"]
["code"]
```
To account for multiple possible captures in each match.

I am currently using the `eachmatch` function as an alternative.

### PR DESCRIPTION
**Pull request by nalimilan**
It is not obvious that returning matching `SubString`s rather than `RegexMatch` objects is a good idea.

Fixes https://github.com/JuliaLang/julia/issues/26049.

### CODE REVIEW

File: stdlib/REPL/src/REPLCompletions.jl [hunk 1]
```diff
@@ -201,7 +201,7 @@ function complete_path(path::AbstractString, pos; use_envpath=false)
     end
 
     matchList = String[replace(s, r"\s" => "\\ ") for s in matches]
-    startpos = pos - lastindex(prefix) + 1 - length(matchall(r" ", prefix))
+    startpos = pos - lastindex(prefix) + 1 - count(c -> c == ' ', prefix)
```
Comment (nalimilan) [inline stdlib/REPL/src/REPLCompletions.jl]: I hope this change is correct. AFAICT it is, but the original code is such a convoluted way of doing this that I feel like I may be missing something.

Comment (mbauman) [reply stdlib/REPL/src/REPLCompletions.jl]: Looks right to me, too.  Could be `equalto(' ')`.

Comment (nalimilan) [reply stdlib/REPL/src/REPLCompletions.jl]: Yeah, I realized that right after pushing. Fixed while rebasing.

**Comment by nalimilan** [comment]
I'll merge once CI passes barring objections.

File: test/regex.jl [hunk 2]
```diff
@@ -4,22 +4,20 @@ function collect_eachmatch(re, str; overlap=false)
...
-    @test f(r"", "") == [""]
-    @test f(r"", "", overlap=true) == [""]
-    @test f(r"aa", "aaaa") == ["aa", "aa"]
-    @test f(r"aa", "aaaa", overlap=true) == ["aa", "aa", "aa"]
-    @test f(r"", "aaa") == ["", "", "", ""]
-    @test f(r"", "aaa", overlap=true) == ["", "", "", ""]
-    @test f(r"GCG","GCGCG") == ["GCG"]
-    @test f(r"GCG","GCGCG",overlap=true) == ["GCG","GCG"]
-end
+@test collect_eachmatch(r"a?b?", "asbd") == ["a","","b","",""] == f(r"""a?b?""", "asbd")
```
Comment (mbauman) [inline test/regex.jl]: You've got an old `f()` call in here.

Comment (nalimilan) [reply test/regex.jl]: Thanks, fixed!

**Review by mbauman — APPROVED** [review]

**Comment by diam** [comment]
Ruby has a `scan` String method for the julia old `matchall` expected behavior.
[https://ruby-doc.org/core-2.4.0/String.html#method-i-scan](https://ruby-doc.org/core-2.4.0/String.html#method-i-scan)

Is it too late to add (or reserve) this name for julia-07?

    words = collect(m.match for m in eachmatch(r"\d+(\.\d+)?", txt))

I'd much more prefer:

    words = scan(r"\d+(\.\d+)?", txt)

-- Maurice

**Comment by StefanKarpinski** [comment]
There's no need to reserve names: exporting new names from Base is not considered breaking.

**Comment by diam** [comment]
Bonjour,

So what is the best way from julia-1.3+ for getting a  String Vector from a regex like with matchall?

```
pat = r"[+-]?\d+\.?\d*"
txt = "..."

# findall?
ranges = findall(pat, txt)
words = map(range->txt[range], ranges) # could be shorter?
numbers = parse.(Float64, words)

# eachmatch?
matches = eachmatch(pat, txt)
words = getfield.(matches, :match)
numbers = parse.(Float64, words)

# versus matchall or getall or... (;-)
numbers = parse.(Float64, matchall(pat, txt))
```
-- Maurice

**Comment by KristofferC** [comment]
Instead of commenting on old issues, it is better to open a new thread at https://discourse.julialang.org/ for questions.

**Comment by kubark42** [comment]
> Instead of commenting on old issues, it is better to open a new thread at https://discourse.julialang.org/ for questions.

I'd like to lightly disagree with this statement. I found this issue by using the changelog. I was looking for why `matchall()` no longer worked, and to understand what its replacement was. Without @diam's question and subsequent link, I would never have found the discourse page.

Of course, the true value is the cross-link to the discourse page, but since that doesn't happen automatically it's valuable to see discourse open here before ultimately moving to a better home.

**Pull request merged**